samples/nrf9160/gps/src/supl_support.c

```diff
@@ -7,108 +7,79 @@
 #include <zephyr.h>
 #include <supl_session.h>
 #include <stdio.h>
-#include <logging/log.h>
 #include <net/socket.h>
 
 #define SUPL_SERVER        "supl.google.com"
 #define SUPL_SERVER_PORT   7276
 
-/* Number of getaddrinfo attempts */
-#define GAI_ATTEMPT_COUNT  3
-
-static int supl_fd;
+static int supl_fd = -1;
 
 int open_supl_socket(void)
 {
-	int err = -1;
-	int proto;
-	int gai_cnt = 0;
-	uint16_t port;
-	struct addrinfo *addr;
+	int err;
 	struct addrinfo *info;
 
-	proto = IPPROTO_TCP;
-	port = htons(SUPL_SERVER_PORT);
-
 	struct addrinfo hints = {
-		.ai_family = AF_INET,
-		.ai_socktype = SOCK_STREAM,
-		.ai_protocol = proto,
-		/* Either a valid,
-		 * NULL-terminated access point name or NULL.
-		 */
-		.ai_canonname = NULL,
-	};
-
-	/* Try getaddrinfo many times, sleep a bit between retries */
-	do {
-		err = getaddrinfo(SUPL_SERVER, NULL, &hints, &info);
-		gai_cnt++;
-
-		if (err) {
-			if (gai_cnt < GAI_ATTEMPT_COUNT) {
-				/* Sleep between retries */
-				k_sleep(K_MSEC(1000 * gai_cnt));
-			} else {
-				/* Return if no success after many retries */
-				printk("Failed to resolve hostname %s on IPv4, errno: %d)\n",
-					SUPL_SERVER, errno);
-
-				return -1;
-			}
-		}
-	} while (err);
-
-	/* Create socket */
-	supl_fd = socket(AF_INET, SOCK_STREAM, proto);
-	if (supl_fd < 0) {
-		printk("Failed to create socket, errno %d\n", errno);
-		goto cleanup;
-	}
-
-	struct timeval timeout = {
-		.tv_sec = 1,
-		.tv_usec = 0,
+		.ai_family = AF_UNSPEC, /* Both IPv4 and IPv6 addresses accepted. */
+		.ai_socktype = SOCK_STREAM
 	};
 
-	err = setsockopt(supl_fd,
-			 SOL_SOCKET,
-			 SO_RCVTIMEO,
-			 &timeout,
-			 sizeof(timeout));
+	err = getaddrinfo(SUPL_SERVER, NULL, &hints, &info);
 	if (err) {
-		printk("Failed to setup socket timeout, errno %d\n", errno);
+		printk("Failed to resolve hostname %s, error: %d\n", SUPL_SERVER, err);
+
 		return -1;
 	}
 
-	/* Not connected */
+	/* Not connected. */
 	err = -1;
 
-	for (addr = info; addr != NULL; addr = addr->ai_next) {
-		char ip[INET_ADDRSTRLEN] = { 0 };
+	for (struct addrinfo *addr = info; addr != NULL; addr = addr->ai_next) {
+		char ip[INET6_ADDRSTRLEN] = { 0 };
 		struct sockaddr *const sa = addr->ai_addr;
 
-		if (sa->sa_family != AF_INET) {
-			/* Not an IPv4 address, try the next address */
-			continue;
+		switch (sa->sa_family) {
```

Comment:
You could provide port number as a `service` string to `getaddrinfo()`, it should set the port for you, no need for this `switch` then.

Reply:
Nice, thanks for the tip.

While testing this change, I noticed that there's a problem with with IPv4v6 PDN contexts. The sample waits for home/roaming network indication from LTE LC and when LTE is connected, it triggers the getaddrinfo() call. However, it seems that immediately after modem has indicated being connected, IPv6 is not yet up. Because of this an IPv4 address is always returned with an IPv4v6 context. I don't know a good way around this. The PDN library does have an event for IPv6 activation, but we don't know if we should wait for that or not. A simple 1s sleep before the getaddrinfo() does fix this, but it's not very nice.